{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bc59523-234e-4185-932d-968daec452f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from reportlab.lib import colors\n",
    "from reportlab.lib.units import cm\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "import pandas as pd\n",
    "from pdf2image import convert_from_path\n",
    "from PIL import ImageOps, Image\n",
    "from ete3 import Tree, TreeNode, TreeStyle, NodeStyle, faces, RectFace"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2da6b00e-d245-46e6-827e-0f6e8af7c8ba",
   "metadata": {},
   "source": [
    "## Generate lists of target locus ID's \n",
    "This pulls from the strictly defined pyl user list (i.e. a genome with complete pylRS, pylT, and pylBCD), and identifies locus ID's for  HMM hits of interest. For the biosynthetic machinery, we're using pylB as a representative since the majority of the pyl biosynethic machinery occurs in an operon. For those genomes where pylB is located far away from pylC and pylD, we manually search and generate the neighborhood."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df8fd003-db66-41c4-b935-ccbe9bf873c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/katharineshalvarjian/Documents/gtdb/3-notebooks'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e474054d-5f48-4eb7-9bc3-ffd318fdc067",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create score dictionary for each HMM\n",
    "score_dict = {}\n",
    "\n",
    "for file in os.listdir(\"../1-pyl/pyl/pyl-hmms/\"):\n",
    "    if \".DS_Store\" in file:\n",
    "        continue\n",
    "    with open(f\"../1-pyl/pyl/pyl-hmms/{file}\", 'r') as f: \n",
    "        for line in f: \n",
    "            if line.startswith('TC'): \n",
    "                gene = file.split('.')[0]\n",
    "                score = float(line.strip().split( )[1])\n",
    "                score_dict.update({gene:score})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9349be7f-0c50-4884-b596-6c8cfece60a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"../analysis_files/2025-02-11_pyl_STRICT.csv\")\n",
    "genomes = df['genome'].tolist()\n",
    "\n",
    "# make loci lists \n",
    "pylB_loci = []\n",
    "pylS_loci = []\n",
    "pylSn_loci = []\n",
    "pylSc_loci = []\n",
    "\n",
    "genes = ['pylB', 'pylS', 'pylSn', 'pylSc']\n",
    "\n",
    "for genome in genomes:\n",
    "    genome_df = pd.read_csv(f\"../1-pyl/pyl/pyl-hmm_hits/{genome}_hmm_hits.csv\", sep=\",\")\n",
    "    \n",
    "    for gene in genes:\n",
    "        score = score_dict[gene]\n",
    "        filtered_df = genome_df[(genome_df['hmm'] == gene) & (genome_df['score'] >= score)]\n",
    "        \n",
    "        if not filtered_df.empty:\n",
    "            gene_id = filtered_df.iloc[0]['geneID']\n",
    "            \n",
    "            if gene == 'pylB':\n",
    "                pylB_loci.append(gene_id)\n",
    "            elif gene == 'pylS':\n",
    "                pylS_loci.append(gene_id)\n",
    "            elif gene == 'pylSc':\n",
    "                pylSc_loci.append(gene_id)\n",
    "            elif gene == 'pylSn':\n",
    "                pylSn_loci.append(gene_id)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d755f5f6-f121-4126-96c5-97f2bc2c229d",
   "metadata": {},
   "source": [
    "## Make an empty diagram\n",
    "Generates an empty/blank .pdf genome diagram to use for nodes that lack components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b22ed321-4360-480f-ac4b-65ecafd1c765",
   "metadata": {},
   "outputs": [],
   "source": [
    "gd_diagram = GenomeDiagram.Diagram()\n",
    "gd_track_for_features = gd_diagram.new_track(1, tracklines=0)\n",
    "gd_feature_set = gd_track_for_features.new_set()\n",
    "\n",
    "gd_diagram.draw(format=\"linear\", orientation=\"landscape\", fragments=1, start=0, end=1, pagesize=(30*cm, 7.5*cm))\n",
    "gd_diagram.write(\"./dummy.pdf\", \"PDF\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f1226bb-6f2a-4b54-93fa-6f49058c3a4a",
   "metadata": {},
   "source": [
    "## Generate neighborhood images \n",
    "Generates genomic neighborhood images (in .pdf format) for each target gene and all genes within a genomic window around that target gene. Note that you should be able to export these with a .png extension, but BioPython was having difficulty loading a PIL sub-package, so I used a pdf > png converter work-around. \n",
    "This happens in several steps: \n",
    "1. define a get_features functions, which returns all features within the specified window length. This takes in the record (a contig), the target feature ID, and a window length and returns all features within that window. Note that because genbank files re-index contigs at 1, isolating the record is essential (otherwise it will an excess of features within the window).\n",
    "2. define a pad_sequence function, which pads sequences with N's when the window length exceeds the contig length. This takes a record (contig), target start, and target end.\n",
    "3. run these functions within a main loop to get features, pad sequences (if neccessary) and generate the output pdf's. Note the 'to_flip' list -- this generates a list of image ID's that we will flip the orientation of (mirror), such that mmp16 is on the forward strand. Because many of these features are on contigs, the assigned orientation is not meaningful.\n",
    "4. convert the pdf's to png's and mirror those images that are listed in 'to_flip'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "70d9b40e-2fa0-4474-8f1d-bd980ca66400",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_features(record, target_feature, window_length):\n",
    "    target_start = max(0, target_feature.location.start - window_length)\n",
    "    target_end = min(len(record.seq), target_feature.location.end + window_length)\n",
    "    features = []\n",
    "\n",
    "    for feature in record.features: \n",
    "        if feature.type == \"CDS\" and feature != target_feature: \n",
    "            if target_start <= feature.location.end and feature.location.start <= target_end: \n",
    "                features.append(feature)\n",
    "        \n",
    "    return features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b08e8baa-6047-4ab6-b0f7-60169974f407",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pad_sequence(record, target_start, target_end):\n",
    "    sequence_length = len(record.seq)\n",
    "    if target_start < 0:\n",
    "        left_padding = abs(target_start)\n",
    "    else:\n",
    "        left_padding = 0\n",
    "\n",
    "    if target_end > sequence_length:\n",
    "        right_padding = target_end - sequence_length\n",
    "    else:\n",
    "        right_padding = 0\n",
    "\n",
    "    padded_sequence = \"N\" * left_padding + str(record.seq) + \"N\" * right_padding\n",
    "    return padded_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3fab7e82-2557-414f-802f-eb4be4e20bbf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/katharineshalvarjian/Documents/gtdb/1-pyl/neighborhood\n"
     ]
    }
   ],
   "source": [
    "cd ../1-pyl/neighborhood/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "13dffadf-d0e0-44a4-90a0-c31c254cb503",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sets EC number for genes of interest\n",
    "pylBCD = ['5.4.99.58', '6.3.2.-', '1.4.1.-']\n",
    "pylS = ['6.1.1.26']\n",
    "mtase = ['2.1.1.247', '2.1.1.248', '2.1.1.249', '2.1.1.250']\n",
    "hyd = ['3.5.3.9']\n",
    "\n",
    "# sets colors for all genes of interest\n",
    "color_map = {\n",
    "    'pylBCD': colors.Color(.118, .533, .898, .8),  # blue\n",
    "    'pylS': colors.Color(1.0, .655, .149, .8),  # orange\n",
    "    'mtase': colors.Color(.804, .863, .224, .8),  # green\n",
    "    'hyd': colors.Color(0.0, .537, .482, .8), # teal\n",
    "    'default': colors.Color(.85, .85, .85, .8) # light gray\n",
    "}\n",
    "\n",
    "genome_directory = '../../archaea_prokka/'\n",
    "\n",
    "# loop through each genome in the list defined as having target gene\n",
    "for genome in genomes:  \n",
    "    # create genome diagram\n",
    "    gd_diagram = GenomeDiagram.Diagram(name=None, tracklines=False)\n",
    "    path = f\"{genome_directory}{genome}/{genome}.gbk\"\n",
    "\n",
    "    # parse genbank records for each genome\n",
    "    records = list(SeqIO.parse(path, \"genbank\"))\n",
    "    target_feature = None\n",
    "    target_record = None\n",
    "\n",
    "    # loop through records and compare to the list of target loci, once found set the contig and feature as found  \n",
    "    for record in records:\n",
    "        for feature in record.features:\n",
    "            if feature.type == \"CDS\" and feature.qualifiers.get('locus_tag', ['N/A'])[0] in pylSc_loci:\n",
    "                target_record = record\n",
    "                target_feature = feature\n",
    "                break\n",
    "\n",
    "    # once target feature is found, set a window length beginning at the start and stop of the target gene          \n",
    "    if target_feature:\n",
    "        window_length = 7500 \n",
    "        target_start = target_feature.location.start - window_length\n",
    "        target_end = target_feature.location.end + window_length\n",
    "\n",
    "        # if window exceeds the contig length, create a padded sequence record & get features\n",
    "        if target_start < 0 or target_end > len(record.seq):\n",
    "            padded_sequence = pad_sequence(target_record, target_start, target_end)\n",
    "            padded_record = SeqIO.SeqRecord(\n",
    "                Seq(padded_sequence),\n",
    "                id=target_record.id,\n",
    "                name=target_record.name,\n",
    "                description=target_record.description\n",
    "            )\n",
    "            features = get_features(target_record, target_feature, window_length)\n",
    "            padded_record.features = features\n",
    "        # otherwise, get features\n",
    "        else:\n",
    "            features = get_features(target_record, target_feature, window_length)\n",
    "\n",
    "        # once features are found, create and fill a new track\n",
    "        if features:\n",
    "            gd_track = gd_diagram.new_track(1, name=\"Features\", start=target_start, end=target_end, hide=False, greytrack=False, scale=False)\n",
    "            gd_feature_set = gd_track.new_set()\n",
    "\n",
    "            pylBCD_features = []\n",
    "            pylS_features = []\n",
    "            mtase_features = []\n",
    "            hyd_features = []\n",
    "            other_features = []\n",
    "\n",
    "            # define target features for track \n",
    "            for feature in features:\n",
    "                if feature != target_feature:\n",
    "                    locus_tag = feature.qualifiers.get('locus_tag', ['N/A'])[0]\n",
    "                    ec_number = feature.qualifiers.get('EC_number', ['N/A'])[0]\n",
    "                    gene = feature.qualifiers.get('gene', ['N/A'])[0]\n",
    "                    product = feature.qualifiers.get('product', ['N/A'])[0]\n",
    "\n",
    "                    if ec_number in pylBCD:\n",
    "                        pylBCD_features.append((feature, gene))\n",
    "                    elif ec_number in pylS:\n",
    "                        pylS_features.append((feature, gene))\n",
    "                    elif ec_number in mtase:\n",
    "                        mtase_features.append((feature, gene))\n",
    "                    elif ec_number in hyd: \n",
    "                        hyd_features.append((feature, ec_number))\n",
    "                    else:\n",
    "                        other_features.append((feature, product))\n",
    "\n",
    "            for feature, gene in pylBCD_features:\n",
    "                gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                                           color=color_map['pylBCD'],\n",
    "                                           label=False, label_size=8, name=gene, arrowshaft_height=1.0)\n",
    "\n",
    "            for feature, gene in pylS_features:\n",
    "                gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                                           color=color_map['pylS'],\n",
    "                                           label=False, label_size=8, name=gene, arrowshaft_height=1.0)\n",
    "\n",
    "            for feature, gene in mtase_features:\n",
    "                gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                                           color=color_map['mtase'],\n",
    "                                           label=False, label_size=8, name=gene, arrowshaft_height=1.0)\n",
    "\n",
    "            for feature, gene in hyd_features:\n",
    "                gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                                           color=color_map['hyd'],\n",
    "                                           label=False, label_size=8, name=gene, arrowshaft_height=1.0)\n",
    "\n",
    "            for feature, ec_number in other_features:\n",
    "                gd_feature_set.add_feature(feature, sigil=\"ARROW\",\n",
    "                                           color=color_map['default'],\n",
    "                                           label=False, label_size=8, name=product, arrowshaft_height=1.0)\n",
    "\n",
    "            gd_feature_set.add_feature(target_feature, sigil=\"ARROW\",\n",
    "                                       color=color_map['pylBCD'],\n",
    "                                       label=False, label_size=8, name='pylB', arrowshaft_height=1.0)\n",
    "\n",
    "            # set page size, draw and write diagram\n",
    "            pagesize = (30 * cm, 7.5 * cm)\n",
    "            gd_diagram.draw(format=\"linear\", orientation=\"landscape\",\n",
    "                            fragments=1, start=target_start, end=target_end, pagesize=pagesize, tracklines=False)\n",
    "            gd_diagram.write(f\"./graphics-pylSc_test/{genome}.pdf\", \"pdf\")\n",
    "\n",
    "            # if the target feature needs to be flipped, flag for flipping with converted to .png\n",
    "            # if target_feature.location.strand == -1: \n",
    "            #     to_flip.append(f\"{genome}.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2fd1150f-3953-4841-b7d0-7640297a8916",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pdf2image import convert_from_path\n",
    "\n",
    "def convert_pdf_to_png(pdf_folder, png_folder):\n",
    "    os.makedirs(png_folder, exist_ok=True)\n",
    "\n",
    "    for pdf_file in os.listdir(pdf_folder):\n",
    "        if pdf_file.endswith(\".pdf\"):\n",
    "            pdf_path = os.path.join(pdf_folder, pdf_file)\n",
    "            images = convert_from_path(pdf_path, dpi=300)\n",
    "\n",
    "            for i, image in enumerate(images):\n",
    "                # Save the image as a PNG\n",
    "                png_file = pdf_file.replace(\".pdf\", \".png\")  # Adding index for multi-page PDFs\n",
    "                png_path = os.path.join(png_folder, png_file)\n",
    "                image.save(png_path, \"PNG\", quality=95, optimize=True)\n",
    "\n",
    "convert_pdf_to_png(\"./graphics-pylB_nolabel/\", \"./png-pylB_nolabel/\")\n",
    "# convert_pdf_to_png(\"./graphics-pylS/\", \"./png-pylS/\")\n",
    "# convert_pdf_to_png(\"./graphics-15000-2/\", \"./png-15000-2/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a15feb70-f1ce-4067-b79f-f80ac67af652",
   "metadata": {},
   "source": [
    "## Make the tree\n",
    "The functions and loops below generate the tree with gene neighborhood images and presence/absence matrix information overlaid. First, we define a layout function that is applied to nodes using ete3, then we generate the tree itself, loading in information as dataframes and dictionaries. Please note that to use this function, you'll need to define a tax_dict that converts between the GTDB accession (sans first 4 characters, i.e. in 'GCF_' or 'GCA_' format) and the species name. You will also have to define a presence_absence_dict. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bacc5abe-f4b4-48c3-86be-c8ac33123264",
   "metadata": {},
   "outputs": [],
   "source": [
    "def layout(node):\n",
    "    # checks if node is a leaf\n",
    "    if node.is_leaf():\n",
    "        try:\n",
    "            # pull node name, convert to species name, add a text_face, & remove the original node ID\n",
    "            node_name = str(node.name[3:])\n",
    "\n",
    "            if node.name in tax_dict:\n",
    "                label = tax_dict[node.name]\n",
    "                node_label = label.split(\";s__\")[1]\n",
    "                text_face = faces.TextFace(f\"{node_label}   ({node_name})\", fsize=125)\n",
    "                text_face.margin_left = 5\n",
    "                text_face.margin_bottom = 10\n",
    "                faces.add_face_to_node(text_face, node, column=0, position=\"aligned\")\n",
    "\n",
    "\n",
    "            img_path = f\"./png-pylB_nolabel/{node_name}.png\"\n",
    "            img_face = faces.ImgFace(img_path, width=1600, height=300)\n",
    "            img_face.margin_left = 30\n",
    "            faces.add_face_to_node(img_face, node, column=1, position=\"aligned\")\n",
    "\n",
    "\n",
    "    # format line widths and colors for each node\n",
    "    nstyle = NodeStyle()\n",
    "    nstyle[\"fgcolor\"] = \"#ffffff\"\n",
    "    nstyle[\"size\"] = 0\n",
    "    nstyle[\"vt_line_color\"] = \"#000000\"\n",
    "    nstyle[\"vt_line_width\"] = 10\n",
    "    nstyle[\"hz_line_color\"] = \"#000000\"\n",
    "    nstyle[\"hz_line_width\"] = 10\n",
    "\n",
    "    node.set_style(nstyle)\n",
    "\n",
    "# make a taxonomy converter\n",
    "df2 = pd.read_csv(\"../../archaea_info/ar53_metadata_tax.csv\")\n",
    "df2[\"species\"] = df2[\"gtdb_taxonomy\"].str.split(\";s__\", expand=True)[1]\n",
    "tax_dict = dict(zip(df2[\"gtdb_accession\"], df2[\"gtdb_taxonomy\"]))\n",
    "\n",
    "# load tree\n",
    "tree = Tree(\"./2025-02-05_pylBCD_checkM95-strict.newick\", format=1)\n",
    "\n",
    "# apply tree style to the parsed tree\n",
    "ts = TreeStyle()\n",
    "ts.layout_fn = layout\n",
    "ts.scale = 5000.0\n",
    "ts.force_topology = False\n",
    "ts.draw_guiding_lines = True\n",
    "ts.margin_left = 1000\n",
    "ts.margin_right = 1000\n",
    "ts.margin_top = 1000\n",
    "ts.margin_bottom = 1000\n",
    "\n",
    "# render tree\n",
    "tree.render(\"./neighborhood-7500.pdf\", tree_style=ts)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
