Scripts and code used in the data analysis and exploratory bioinformatics for Pyrrolysine genomics and physiology in methanogens. You can find the pre-print describing the outcomes of this work [here](insert link!). Descriptions for notebooks and scripts are below:
  * **pyl-concatenate.ipynb**: notebook used to concated pylRS and pylBCD sequence alignments for tree and tanglegram building.
  * **pyl-neighborhood_diagram.ipynb**: notebook used to generate neighborhood diagrams for pylB (+/- 7.5 kB)
  * **pyl-transcript.ipynb**: notebook used to generate Pyl supply and Pyl demand estimates based on FPKM values generated in RNA-Seq.
  * **pyl_id.py**: script used to generate predictions for TAG-containing genes based on prokka annotations.
  * **pyl_id_ncbi.py**: script used to generate predictions for TAG-containing genes based on the NCBI genbank format.
